File: crates/agent/src/stages.rs
//! THE LOOP A TURN RUNS — and, since the strategy stage, the loop the turn
//! CHOOSES.
//!
//! An agent file used to name its stages outright (`stages: [plan, work,
//! verify]`) and every turn walked all of them. That made the loop a property
//! of the agent when it is a property of the message: the same assistant that
//! should answer "hello" in one call should plan, work, check and critique when
//! asked to build something. `strategy` is one cheap call that reads the
//! message and picks; `crate::strategy` holds the three routes and why an
//! unreadable vote lands in the middle one.
//!
//! A STAGE IS NOT A NEW MACHINE. It is one instruction pushed into the paper
//! and one more call, taken by the same `step` against the same window: the
//! stage's prose reply, instead of ending the turn, moves the cursor on and
//! opens the next one. So a stage cannot invent a transition the loop did not
//! already have, and there is no second state machine to keep in agreement
//! with the first.
//!
//! WHY NOT TOOLS IN EVERY STAGE. `strategy`, `plan`, `critique` and `answer`
//! may not act, and `ask::scoped_tools` enforces it rather than trusting the
//! sentence — the `engine: base` lesson (19): a capability that is described
//! but not enforced is a setting that looks applied. `plan` is the one
//! exception and a narrow one: it is granted the two skill tools, because its
//! brief tells it to read skills and an instruction to call a tool the agent
//! was never given is noise.

use kernel::{EventKind, Timestamp};

use crate::ask::call_model;
use crate::brief;
use crate::effect::Effect;
use crate::state::AgentState;
use crate::strategy::{self, Route, STRATEGY};

pub const PLAN: &str = "plan";
pub const WORK: &str = "work";
pub const VERIFY: &str = "verify";
pub const CRITIQUE: &str = "critique";
/// Work with the tools taken away, which is what the `answer` route is. A
/// separate name rather than a flag on `work` because `stages::tools_on` is
/// read in four places and a stage list is the thing they all read.
pub const ANSWER: &str = "answer";

/// The stages that exist — `spec` refuses any other name for the reason
/// `engine` refuses `reakt`: a stage that parses clean and selects nothing is
/// worse than no stage key at all.
pub const STAGES: [&str; 6] = [STRATEGY, PLAN, WORK, VERIFY, CRITIQUE, ANSWER];

/// The fact a stage was entered. Emitted for `verify::VERIFY_NUDGED`'s reason
/// (I8): the machine added a round, and a round nobody can see is a model
/// talking to itself while the token meter charges for it.
pub const STAGE_ENTERED: &str = "core.stage_entered";

/// The fact a route was chosen, and what decided it. The strategy stage spends
/// a call to make a decision the person never sees otherwise; a turn that
/// silently became four calls instead of one is the sort of thing a bill
/// explains and nothing else does.
pub const ROUTE_CHOSEN: &str = "core.route_chosen";

pub fn is_stage(name: &str) -> bool {
    STAGES.contains(&name)
}

/// Whether this stage may call the agent's full toolbox.
pub fn tools_on(stage: &str) -> bool {
    brief::acts(stage)
}

/// The stage the turn is in. Absent `stages:` reads as `work`, which is what
/// makes a file with no stage key behave exactly as it did before this existed.
pub(crate) fn current(state: &AgentState) -> &str {
    state.stages.get(state.stage).map(String::as_str).unwrap_or(WORK)
}

/// Enter the stage the cursor is on: its instruction into the window, and the
/// fact into the log.
pub(crate) fn enter(state: &mut AgentState, at: Timestamp) -> Effect {
    let stage = current(state).to_string();
    let mut said = brief::brief(&stage).to_string();
    if stage == PLAN && state.space.is_some() {
        said.push_str(brief::DURABLE);
    }
    // Into its own block, not into the conversation. A stage with no brief
    // writes an empty one, which is how the block disappears on `work` rather
    // than lingering with the previous stage's instruction still in it.
    let directive = crate::components::Directive { text: said };
    crate::paper::set_component(&mut state.paper, &directive, at);
    entered(&stage)
}

/// The start of a turn: the strategy vote first, then whatever it chooses.
///
/// A file that declares its own `stages:` keeps them — the route machinery is
/// what an agent gets by declaring `[strategy]`, not something imposed on every
/// file. Declaring nothing is still the bare react loop.
pub(crate) fn open(state: &mut AgentState, at: Timestamp) -> Vec<Effect> {
    state.stage = 0;
    crate::passes::open(state);
    // A turn that ran a route last time starts over from the declaration, or
    // the second message of a conversation would inherit the first's plan.
    state.stages = state.declared.clone();
    match state.stages.is_empty() {
        true => Vec::new(),
        false => vec![enter(state, at)],
    }
}

/// A stage produced prose. `None` when that was the last one AND the turn has
/// not earned another pass — then the turn ends the way it always has, through
/// `ending`.
pub(crate) fn next(state: &mut AgentState, reply: &str, at: Timestamp) -> Option<Vec<Effect>> {
    // THE VOTE IS THE ONLY REPLY THAT REWRITES THE LIST. Read here rather than
    // in `step` because this is where a stage's reply is already in hand, and
    // splitting "read the reply" from "act on it" is how the two drift.
    if current(state) == STRATEGY {
        return Some(route(state, reply, at));
    }
    if state.stage + 1 >= state.stages.len() {
        // …OR ROUND AGAIN, back to `work` and never to the start (22). The
        // budget and the mechanical continue condition are both in `passes`.
        return crate::passes::again(state, at);
    }
    state.stage += 1;
    // The evidence flags are NOT cleared here: they are the turn's, and a
    // verify stage that forgot what the work stage wrote would ask the model
    // to check nothing (`verify.rs` — order is the freshness rule).
    let entered = enter(state, at);
    let call = call_model(state, at);
    Some(vec![entered, call])
}

/// Install the route the vote named and open its first stage. The vote itself
/// never reaches the person: it is a decision about the turn, not a turn.
fn route(state: &mut AgentState, reply: &str, at: Timestamp) -> Vec<Effect> {
    let route = strategy::route_of(reply);
    state.stages = route.stages();
    state.stage = 0;
    vec![chosen(route, reply), enter(state, at), call_model(state, at)]
}

/// Whether a `verify` stage is still AHEAD in this turn. The mechanical gate
/// in `verify.rs` and a declared verify stage answer the same question — "a
/// file changed and nothing has run since" — and a turn that fired both asked
/// the model twice and printed two notices saying it (browser walk, 20). The
/// declaration wins: it is the loop this agent's own file asked for.
pub(crate) fn verify_ahead(state: &AgentState) -> bool {
    state.stages.iter().skip(state.stage + 1).any(|s| s == VERIFY)
}

pub(crate) fn entered(stage: &str) -> Effect {
    Effect::Emit {
        kind: EventKind::Custom {
            kind: STAGE_ENTERED.into(),
            payload_json: serde_json::to_string(stage).unwrap_or_else(|_| "\"\"".into()),
        },
    }
}

fn chosen(route: Route, reply: &str) -> Effect {
    // The WHY line, if the model wrote one — the difference between "the
    // machine chose project" and "the machine chose project because the
    // message asked for a working script".
    let why = reply
        .lines()
        .find_map(|l| l.trim().strip_prefix("WHY:"))
        .unwrap_or_default()
        .trim();
    Effect::Emit {
        kind: EventKind::Custom {
            kind: ROUTE_CHOSEN.into(),
            payload_json: serde_json::json!({ "route": route.as_str(), "why": why }).to_string(),
        },
    }
}

/// Which stage a `STAGE_ENTERED` fact names, for the projections.
///
/// EVERY ASSERTION ABOUT THIS FILE IS IN `tests/stages.rs`, through `step` and
/// against the real shipped agent files — what a stage does is a sequence of
/// effects a turn produces, and a unit test of the cursor could pass while the
/// turn it drives ended in the wrong place.
pub fn stage_of(payload_json: &str) -> String {
    serde_json::from_str::<String>(payload_json).unwrap_or_default()
}

/// Which route a `ROUTE_CHOSEN` fact names, and what decided it.
pub fn route_of(payload_json: &str) -> (String, String) {
    let read = |v: &serde_json::Value, k: &str| {
        v.get(k).and_then(|s| s.as_str()).unwrap_or_default().to_string()
    };
    serde_json::from_str::<serde_json::Value>(payload_json)
        .map(|v| (read(&v, "route"), read(&v, "why")))
        .unwrap_or_default()
}

File: packages/agent/src/stages.js
/**
 * A STAGE: what the model is told this pass, what it may call, and the shape
 * the reply must take. An agent file's `stages:` list is the loop it walks.
 *
 * IN RUST THERE WAS A PHASE MACHINE BESIDE THIS, and it is retired. `v1_phases()`
 * returned a table looked up by `AgentState.phase`; the table had one row,
 * `phase` was assigned nowhere outside the constructor, and the `exits` table
 * had a write and no reader. The name survives only in this paragraph, which is
 * the reason a rename would erase: a machine that never had a second state is
 * not a machine, and porting one would be porting the intention.
 *
 * A STAGE IS NOT A NEW MACHINE EITHER. It is one instruction pushed into the
 * paper and one more call, taken by the same `step` against the same window:
 * the stage's prose reply, instead of ending the turn, moves the cursor on. So
 * a stage cannot invent a transition the loop did not already have.
 *
 * WHY MOST STAGES MAY NOT ACT. `strategy`, `critique` and `answer` are refused
 * tools by their ALLOWLIST rather than by their brief's sentence — the
 * `engine: base` lesson: a capability described and not enforced is a setting
 * that looks applied. `plan` is the one narrow exception; see below.
 * @module
 */

import { ANSWER, STRATEGY_SCHEMA } from './strategy.js'

/** @typedef {import('@harness/kernel').ToolId} ToolId */
/** @typedef {import('./strategy.js').StageName} StageName */
/** @typedef {import('./frontmatter.js').Refusal} Refusal */

/**
 * Which of an agent's tools a stage exposes. `none` is structural, not a
 * refusal at call time: the granted set is what the affordances section is
 * built from, so a stage that may not act cannot even NAME a tool to the model.
 * @typedef {{kind: 'none'} | {kind: 'all'} | {kind: 'only', tools: ToolId[]}} ToolScope
 */

/** @type {ToolScope} */
export const NO_TOOLS = Object.freeze({ kind: 'none' })

/** Everything THIS agent's file gave it. The per-agent decision belongs to the frontmatter's `tools:` key; a stage says only whether it may act. @type {ToolScope} */
export const ALL_TOOLS = Object.freeze({ kind: 'all' })

/** @param {readonly ToolId[]} tools @returns {ToolScope} */
export function onlyTools(tools) {
  return { kind: 'only', tools: [...tools] }
}

/**
 * The scope applied to one agent's toolbox — the ONLY narrowing there is.
 * Generic over the element rather than over a toolbox type because the whole of
 * the operation is a filter by name. Order is the TOOLBOX's, never the scope's:
 * the model reads these as lines, and the agent file decided their order.
 * @template {{name: string}} T
 * @param {ToolScope} scope @param {readonly T[]} tools @returns {T[]}
 */
export function grant(scope, tools) {
  if (scope.kind === 'none') return []
  if (scope.kind === 'all') return [...tools]
  return tools.filter((tool) => scope.tools.includes(tool.name))
}

/**
 * The exact reply shape demanded back — parsed, never trusted. TWO, DOWN FROM
 * FOUR: `PlanSteps` and `Verdict` had zero construction sites between them.
 * @typedef {'tool_envelope' | 'answer'} ResponseContract
 */

/** @type {readonly ResponseContract[]} */
export const RESPONSE_CONTRACTS = /** @type {const} */ (['tool_envelope', 'answer'])

/**
 * WHAT ONE WORKING TURN'S PAPER MAY COST. **PROVISIONAL.** Measured 2026-08-23:
 * the shipped `main` wanted 4174 tokens before a conversation existed against a
 * budget of 4096, so the ladder elided `## observations` on every turn — while
 * the agent's own prose told it to read that block. 8192 leaves ~4k for the
 * window beside a ~4.2k standing paper and keeps the whole prompt inside the
 * 16k context of the local models this page is pointed at. A number chosen
 * against a MEASURED model context would be better, and that measurement is the
 * owner's, which is why this is provisional.
 * @type {{maxTokens: number}}
 */
export const WORK_BUDGET = Object.freeze({ maxTokens: 8192 })

/**
 * One stage, resolved: the words it enters with, what it may call, and the
 * shape it must answer in. `responseSchema` is null for every stage whose reply
 * only a PERSON reads — only a reply the MACHINE parses needs to override what
 * the loop would have asked for anyway, and today exactly one does.
 * @typedef {{name: StageName, brief: string, toolAllowlist: ToolScope, responseSchema: typeof STRATEGY_SCHEMA | null}} Stage
 */

/** The two tools the `plan` stage is granted. Its brief tells it to read skills; refusing it every tool would make that instruction a lie, and granting the whole toolbox would let it start the work it is supposed to be planning. @type {ToolId[]} */
export const SKILL_TOOLS = ['list_skills', 'read_skill']

/** @type {Record<StageName, ToolScope>} */
const ALLOWLISTS = {
  work: ALL_TOOLS,
  verify: ALL_TOOLS,
  plan: onlyTools(SKILL_TOOLS),
  strategy: NO_TOOLS,
  critique: NO_TOOLS,
  [ANSWER]: NO_TOOLS,
}

/**
 * Every brief there is. `work` and `answer` are absent BY DESIGN — the person's
 * own request is the instruction there, and a second one would compete with it.
 * `durable` is not a stage: it is the paragraph appended to `plan` for an agent
 * that has a space, a key rather than the tail of `plan.md`, because the
 * alternative is core splitting one file on a separator, which is parsing a
 * brief.
 * @type {readonly string[]}
 */
export const BRIEF_KEYS = /** @type {const} */ (['strategy', 'plan', 'verify', 'critique', 'durable'])

export const DURABLE = 'durable'

/** The file a key is read from, built here so the load refusal and the stage refusal cannot name two different paths. @param {string} key @returns {string} */
export function briefPath(key) {
  return `public/stages/${key}.md`
}

/**
 * EVERY BRIEF, OR A REFUSAL. An unknown key, a missing one, or one blank once
 * trimmed refuses the WHOLE SET rather than the one key: a half-loaded set is
 * an app that runs until the turn that needed the missing one.
 *
 * There is no compiled-in copy to fall back to, deliberately. A default here
 * would be a stage that writes no plan and names no check while looking exactly
 * like a stage that ran.
 * @param {Iterable<{key: string, text: string}>} files
 * @returns {{briefs: Record<string, string>} | {refusal: Refusal}}
 */
export function loadBriefs(files) {
  /** @type {Record<string, string>} */
  const briefs = {}
  for (const { key, text } of files) {
    if (!BRIEF_KEYS.includes(key)) {
      return { refusal: { path: briefPath(key), key, message: `No stage is briefed by the name "${key}" — the briefs are: ${BRIEF_KEYS.join(', ')}.` } }
    }
    if (text.trim() === '') {
      return { refusal: { path: briefPath(key), key, message: `${briefPath(key)} is empty, and a stage entered with no instruction looks exactly like one that ran.` } }
    }
    briefs[key] = text.trim()
  }
  const missing = BRIEF_KEYS.find((key) => !Object.hasOwn(briefs, key))
  return missing === undefined
    ? { briefs }
    : { refusal: { path: briefPath(missing), key: missing, message: `${briefPath(missing)} was not loaded, so the ${missing} stage has nothing to say.` } }
}

/**
 * ONE STAGE, RESOLVED against the briefs this build loaded. A briefed stage
 * whose file never arrived REFUSES rather than entering empty — that is the
 * whole reason this returns a value instead of a Stage.
 *
 * The `\n\n` before the durable paragraph is the APPENDER's business and never
 * the file's.
 * @param {StageName} name
 * @param {{briefs: Record<string, string>, hasSpace?: boolean}} of
 * @returns {{stage: Stage} | {refusal: Refusal}}
 */
export function resolveStage(name, of) {
  const parts = name === 'plan' && of.hasSpace ? ['plan', DURABLE] : [name]
  /** @type {string[]} */
  const said = []
  for (const key of parts) {
    if (!BRIEF_KEYS.includes(key)) continue
    const text = of.briefs[key]
    if (text === undefined || text === '') {
      return { refusal: { path: briefPath(key), key, message: `The ${name} stage cannot be entered: ${briefPath(key)} never loaded, and a stage with no instruction writes nothing while looking like it worked.` } }
    }
    said.push(text)
  }
  return {
    stage: {
      name,
      brief: said.join('\n\n'),
      toolAllowlist: ALLOWLISTS[name],
      responseSchema: name === 'strategy' ? STRATEGY_SCHEMA : null,
    },
  }
}

/** Whether this stage may call the agent's full toolbox — written as what MAY act, so a stage added to the vocabulary takes nothing by omission (I6). @param {StageName} name @returns {boolean} */
export function actsIn(name) {
  return ALLOWLISTS[name].kind === 'all'
}
